backend/src/neuroagent/app/dependencies.py

```diff
@@ -415,6 +444,72 @@ async def get_selected_tools(
         return selected_tools
 
 
+async def filtered_tools(
+    request: Request,
+    thread: Annotated[Threads, Depends(get_thread)],
+    tool_list: Annotated[list[type[BaseTool]], Depends(get_selected_tools)],
+    openai_client: Annotated[AsyncOpenAI, Depends(get_openai_client)],
+    settings: Annotated[Settings, Depends(get_settings)],
+) -> list[type[BaseTool]]:
+    """Based on the current conversation, select relevant tools."""
+    if request.method == "GET":
+        return tool_list
+
+    # Awaiting here makes downstream calls already loaded so no performance issue
+    messages = await thread.awaitable_attrs.messages
+    openai_messages = await messages_to_openai_content(messages)
+
+    # Remove the content of tool responses to save tokens
+    for message in openai_messages:
+        if message["role"] == "tool":
+            message["content"] = "..."
+
+    body = await request.json()
+    openai_messages.append({"role": "user", "content": body["content"]})
+
+    system_prompt = f"""TASK: Filter tools for AI agent based on conversation relevance.
+
+AVAILABLE TOOLS:
+{chr(10).join(f"{tool.name}: {tool.description}" for tool in tool_list)}
+
+INSTRUCTIONS:
+1. Analyze the conversation to identify required capabilities
+2. Select at least 5 of the most relevant tools by name only
+3. BIAS TOWARD INCLUSION: If uncertain about a tool's relevance, include it - better to provide too many tools than too few
+4. Only exclude tools that are clearly irrelevant to the conversation
+5. Output format: comma-separated list of tool names
+6. Do not respond to user queries - only filter tools
+
+OUTPUT: [tool_name1, tool_name2, ...]"""
+
+    tool_names = [tool.name for tool in tool_list]
+    TOOL_NAMES_LITERAL = Literal[*tool_names] if len(tool_names) > 0 else str
+
+    class ToolSelection(BaseModel):
+        """Data class for tool selection by an LLM."""
+
+        selected_tools: list[TOOL_NAMES_LITERAL] = Field(  # type: ignore
+            min_length=min(len(tool_names), settings.tools.min_tool_selection),
+            description=f"List of selected tool names, minimum {min(len(tool_names), settings.tools.min_tool_selection)} items. Must contain all of the tools relevant to the conversation.",
+        )
+
+    # Rest of your code remains the same
+    response = await openai_client.beta.chat.completions.parse(
+        messages=[{"role": "system", "content": system_prompt}, *openai_messages],  # type: ignore
+        model="gpt-4o-mini",
+        response_format=ToolSelection,
+    )
+
+    if response.choices[0].message.parsed:
+        selected_tools = response.choices[0].message.parsed.selected_tools
+        logger.info(
+            f"QUERY: {body['content']}, #TOOLS: {len(selected_tools)}, SELECTED TOOLS: {selected_tools}"
+        )
+        return [tool for tool in tool_list if tool.name in selected_tools]
+    else:
+        return []
```

Comment:
IMO this entire structured output classification thing can be put in some helper function. Note a must.